### PR TITLE
Global or local whitening/normalization for refine_template

### LINKED ISSUE #99: Inconsistent cross-correlation values between `match_template` program and `refine_template` program
## Issue

When comparing raw cross-correlation values for the same particles between the Leopard-EM `match_template` and `refine_template` program, there are meaningful differences between the two. The main result of this issue is that program results become hard to quantitatively compare although results within each program are self-consistent.

These differences arise from the `refine_template` program using local information from the extracted particle boxes to calculate the whitening filter while the `match_template` program uses the global power spectra to filter the image.

## Potential fixes

- Only use global whitening within the `ParticleStack` class when calculating the Fourier filters
- Add argument to switch between local/global whitening within the particle stack

Both of these options should be relatively easy to implement within the `ParticleStack` class, and would expect the raw cross-correlation values remain consistent (should test this).

### PR DESCRIPTION
**Pull request by jdickerson95**
Refine template now either fully uses full micrograph for filters and normalization (making it same as match results), or uses the particle for filters and normalization.
The everything is loaded in utils.py and passed into particle stack, which will make it easy to refine from a movie.
Closes #99

### CODE REVIEW

**Review by mgiammar — CHANGES_REQUESTED** [review]

**Review by mgiammar — APPROVED** [review]

File: src/leopard_em/backend/core_refine_template.py [hunk 1]
```diff
@@ -105,6 +106,9 @@ def core_refine_template(
         The number of cross-correlations to process in one batch, defaults to 32.
     num_cuda_streams : int, optional
         Number of CUDA streams to use for parallel processing. Defaults to 1.
+    transform_matrix : torch.Tensor | None, optional
+        Anisotropic magnification transform matrix of shape (2, 2). If None,
+        no magnification transform is applied. Default is None.
```
Comment (mgiammar) [inline src/leopard_em/backend/core_refine_template.py]: Are there anything more general transformations that would fit into here besides anisotropic magnification? If not, renaming this parameter would be a good idea to disambiguate from some real-space transformation like a particle shift, etc.

File: src/leopard_em/backend/cross_correlation.py [hunk 2]
```diff
@@ -226,13 +261,24 @@ def do_batched_orientation_cross_correlate(
...
+            if requires_grad:
+                # Don't use out= parameter to preserve autograd
+                cross_correlation[k, j, ...] = torch.fft.irfftn(
+                    projections_dft, dim=(-2, -1)
+                )
+            else:
+                # Use out= parameter for memory efficiency
+                torch.fft.irfftn(
+                    projections_dft, dim=(-2, -1), out=cross_correlation[k, j, ...]
+                )
```
Comment (mgiammar) [inline src/leopard_em/backend/cross_correlation.py:281]: I forget if the `out=...` parameter here is actually doing anything. The cross-correlation tensor is already allocated and written to. If no difference, can simplify code path.

File: src/leopard_em/utils/ctf_utils.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        Parsed dict if value was a JSON string, original dict if already a dict,
+        or None if value was None or NaN.
+    """
+    # Handle NaN values from empty CSV fields (pandas converts empty fields to NaN)
+    if value is None or (isinstance(value, float) and np.isnan(value)):
+        return None
+    if isinstance(value, str) and value:
+        return json.loads(value)
+    # Already a dict (backward compatibility)
+    return value
```
Comment (mgiammar) [inline src/leopard_em/utils/ctf_utils.py]: Function return value should be `dict | None`, also should include a check to ensure returned value _is_ a dict raising error otherwise.

```
...
if isinstance(value, dict):
    return value

# Raise ValueError (or some other error) indicating the parsing issue
raise ValueError(...)

File: src/leopard_em/utils/ctf_utils.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+                    dtype=torch.float32,
+                )
+
+    # Parse JSON strings for zernike coefficients if they're stored as strings
+    even_zernikes_dict = _parse_json_string_from_series_value(
+        particle_stack["even_zernikes"].iloc[0]
+    )
+    odd_zernikes_dict = _parse_json_string_from_series_value(
+        particle_stack["odd_zernikes"].iloc[0]
+    )
```
Comment (mgiammar) [inline src/leopard_em/utils/ctf_utils.py:269]: Assumes all Zernike components are the same across the entire dataframe (not explicitly checked above). Same happens for the anisotropic magnification matrix. Need to add assertion statements

File: src/leopard_em/utils/ctf_utils.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    mag_matrix_value = particle_stack["mag_matrix"].iloc[0]
+    mag_matrix_tensor = None
+    if mag_matrix_value is not None and not (
+        isinstance(mag_matrix_value, float) and np.isnan(mag_matrix_value)
+    ):
+        # mag_matrix_value might be a list or a string representation of a list
+        if isinstance(mag_matrix_value, str):
+            mag_matrix_list = ast.literal_eval(mag_matrix_value)
+        else:
+            mag_matrix_list = mag_matrix_value
```
Comment (mgiammar) [inline src/leopard_em/utils/ctf_utils.py:249]: Having to leverage `ast.literal_eval` feels like masking an underlying data structuring problem. If value is a string, we know it should _always_ have 4 float values and could use:

```
mag_matrix_list = [float(x) for x in mag_matrix_value.strip("[]").split(",")]
assert len(mag_matrix_list) == 4
```

Or could restructure the CSV file into four columns leaving atomic values for each of the anisotropic magnification components

**Review by mgiammar — CHANGES_REQUESTED** [review]
@jdickerson95 I've gone through the smaller files in the PR and have made some change requests. Still need to go over the larger changes, namely `core_differentiable_refine.py`, `particle_stack.py`, `refine_template_manager.py`, and `backend_setup.py`.

Something I'm noticing with this PR is we're staring to leverage the result CSV file heavily for particle parsing when there are many elements that are hard to fit into a tabular data format (e.g. Zernike dict, aniso mag params). I'm wondering if there's utility in working with JSON data more frequently for describing data results. Open question and not in-scope for this PR.

File: src/leopard_em/backend/core_refine_template.py (hunk 1 above)
Comment (jdickerson95) [reply src/leopard_em/backend/core_refine_template.py]: You could potentially start to model position sensitive distortions such as pin cushion, barrel, and spiral. This is not done by any SPA software since it's only significant if there is something physically wrong with the microscope (e.g. a large magnetic field too close to the microscope). You would also correct images instead of projections so would be done in setup. 

I think renaming it makes sense.

File: src/leopard_em/backend/cross_correlation.py (hunk 2 above)
Comment (jdickerson95) [reply src/leopard_em/backend/cross_correlation.py:281]: I thought a temporary tensor was created whether or not the tensor was pre-allocated, unless out= was specified

Comment (mgiammar) [reply src/leopard_em/backend/cross_correlation.py:281]: Quick profiling shows that indeed using the `out=` speeds correlations up. Resolving.

File: src/leopard_em/backend/core_differentiable_refine.py [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+from leopard_em.utils.ctf_utils import calculate_ctf_filter_stack_full_args
+
+# This is assuming the Euler angles are in the ZYZ intrinsic format
+# AND that the angles are ordered in (phi, theta, psi)
+EULER_ANGLE_FMT = "ZYZ"
+
+
+def combine_euler_angles(angle_a: torch.Tensor, angle_b: torch.Tensor) -> torch.Tensor:
+    """Helper function for composing rotations defined by two sets of Euler angles."""
+    # Ensure both input angles have the same dtype
```
Comment (mgiammar) [inline src/leopard_em/backend/core_differentiable_refine.py]: This function is the same as in the core_refine_template.py file, right? Should move to common utils then

File: src/leopard_em/backend/core_differentiable_refine.py [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+                mag_matrix=mag_matrix,
+            )
+        else:
+            cross_correlation = do_batched_orientation_cross_correlate_cpu(
+                image_dft=particle_image_dft,
+                template_dft=template_dft,
+                rotation_matrices=rot_matrix_batch,
+                projective_filters=combined_projective_filter,
+                mag_matrix=mag_matrix,
+            )
```
Comment (mgiammar) [inline src/leopard_em/backend/core_differentiable_refine.py]: Would the differentiable refine backend function ever be called with a CPU device? In principle, it could work, but not sure if it's been tested or would practically be useful. If not, then should raise a ValueError somewhere up the call stack saying that the differentiable refinement can only be run on a GPU device.

File: src/leopard_em/backend/core_differentiable_refine.py [hunk 8]
```diff
@@ -0,0 +1,712 @@
...
+        The number of cross-correlations to process in one batch, defaults to 32.
+    num_cuda_streams : int, optional
+        Number of CUDA streams to use for parallel processing. Defaults to 1.
+    mag_matrix : torch.Tensor | None, optional
+        Anisotropic magnification matrix of shape (2, 2). If None,
+        no magnification transform is applied. Default is None.
+
+    Returns
+    -------
+    dict[torch.Tensor, torch.Tensor]
```
Comment (mgiammar) [inline src/leopard_em/backend/core_differentiable_refine.py]: Return type should still be string and torch.Tensor dictionary

File: src/leopard_em/backend/core_refine_template.py [hunk 9]
```diff
@@ -393,6 +404,9 @@ def _core_refine_template_single_gpu(
         Torch device to run this process on.
     num_cuda_streams : int, optional
         Number of CUDA streams to use for parallel processing. Defaults to 1.
+    mag_matrix : torch.Tensor | None, optional
+        Anisotropic magnification matrix of shape (2, 2). If None,
+        no magnification transform is applied. Default is None.
     """
```
Comment (mgiammar) [inline src/leopard_em/backend/core_refine_template.py:396]: mag_matrix is missing from the `to(device)` statements

File: src/leopard_em/backend/core_differentiable_refine.py (hunk 6 above)
Comment (jdickerson95) [reply src/leopard_em/backend/core_differentiable_refine.py]: The same with an euler angle dtype check, because you can easily have a dtype mismatch if you are passing euler angles directly into core_differientiable_refine without Leopard-EM backend setup. I can combine to one with the dtype check.

**Review by mgiammar — COMMENTED** [review]
@jdickerson95 Besides the outlying comments, PR looks good to me. Once those are merged and have verification that the code is working for the optimization workflows happy to merge

**Comment by jdickerson95** [comment]
I think I've handled all the changes.
I agree the list/dict in the results csv is getting a bit much, and a json format could be more readable.

I'll test that nothing broke the optimization workflows before merging with dev 1.2

**Review by mgiammar — APPROVED** [review]
